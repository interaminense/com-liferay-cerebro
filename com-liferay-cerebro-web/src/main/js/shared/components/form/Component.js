import autobind from 'autobind-decorator';
import Component, {Config} from 'metal-jsx';

import {checkRequired} from '../../util/form-validators';

class FormComponent extends Component {
	attached() {
		this.on('valueChanged', this.informChange);

		const {ancestorForm} = this.context;

		if (ancestorForm) {
			ancestorForm.addFormElement(this);
		}
	}

	detached() {
		const {ancestorForm} = this.context;

		if (ancestorForm) {
			ancestorForm.removeFormElement(this);
		}
	}

	getFormElementName() {
		return this.props.name;
	}

	getFormElementValue() {
		const {name, processValue} = this.props;

		let {value} = this.state;

		if (processValue) {
			value = processValue(value, name);
		}

		return value;
	}

	@autobind
	informChange() {
		if (this.context.ancestorForm) {
			this.context.ancestorForm.informChange(this.props.name);
		}
	}

	isRequired() {
		const {validator} = this.props;

		return validator && checkRequired(validator);
	}

	validateFormElement() {
		return true;
	}
}

FormComponent.PROPS = {
	name: Config.string().required(),
	processValue: Config.func()
};

FormComponent.STATE = {
	value: Config.any()
};

export {Config} from 'metal-jsx';

export default FormComponent;