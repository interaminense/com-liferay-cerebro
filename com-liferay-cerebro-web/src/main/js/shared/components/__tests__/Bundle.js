import Component from 'metal-jsx';
import Promise from 'metal-promise';

import bundle from '../Bundle';

describe('Bundle', () => {
	let component;

	afterEach(() => {
		if (component) {
			component.dispose();
		}
	});

	class TestComponent extends Component {
		render() {
			return <h1>{'Test Component'}</h1>;
		}
	}

	it('should return a new component', () => {
		const BundledComponent = bundle(() => {
			return Promise.resolve({
				default: TestComponent
			});
		});

		expect(BundledComponent).toBeInstanceOf(Function);
	});

	it('should render a component returned from an import promise', () => {
		const BundledComponent = bundle(() => {
			return Promise.resolve({
				default: TestComponent
			});
		});

		component = new BundledComponent();

		expect(component.element).toBeNull();

		jest.runAllTimers();

		expect(component).toMatchSnapshot();
	});

	it('should statically cache the component on the new class after load', () => {
		const BundledComponent = bundle(() => {
			return Promise.resolve({
				default: TestComponent
			});
		});

		component = new BundledComponent();

		expect(component.element).toBeNull();

		jest.runAllTimers();

		component.dispose();

		jest.runAllTimers();

		component = new BundledComponent();

		expect(component.element).not.toBeNull();
	});
});