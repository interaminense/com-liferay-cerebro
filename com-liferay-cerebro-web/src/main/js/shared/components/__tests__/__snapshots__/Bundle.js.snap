// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Bundle should render a component returned from an import promise 1`] = `<h1>Test Component</h1>`;
